src/tests/cli.cpp

```diff
@@ -601,6 +603,184 @@ TEST_F(rnp_tests, test_cli_rnpkeys)
+                       "1024",
+                       NULL);
+    return ret;
+}
+
+TEST_F(rnp_tests, test_cli_rnpkeys_genkey)
+{
+    assert_false(RNP_MKDIR(GENKEYS, S_IRWXU));
+    time_t   basetime = time(NULL);
+    time_t   rawtime = basetime + 604800;
+    time_t   y2k38time = INT32_MAX;
+    uint32_t expected_diff_beyond2038_absolute;
+    if (rnp_y2k38_warning(y2k38time)) {
+        // we're on the system that doesn't support dates beyond y2k38
+        auto diff_to_y2k38 = y2k38time - basetime;
+        expected_diff_beyond2038_absolute = diff_to_y2k38;
+    } else {
+        time_t    now = time(NULL);
+        struct tm tm2100 = *localtime(&now);
+        tm2100.tm_hour = 0;
+        tm2100.tm_min = 0;
+        tm2100.tm_sec = 0;
+        tm2100.tm_mday = 1;
+        tm2100.tm_mon = 0;
+        tm2100.tm_year = 200;
+        expected_diff_beyond2038_absolute = mktime(&tm2100) - basetime;
+    }
+    struct tm *timeinfo = localtime(&rawtime);
+    // clear hours, minutes and seconds
+    timeinfo->tm_hour = 0;
+    timeinfo->tm_min = 0;
+    timeinfo->tm_sec = 0;
+    rawtime = mktime(timeinfo);
+    auto exp =
+      fmt("%d-%02d-%02d", timeinfo->tm_year + 1900, timeinfo->tm_mon + 1, timeinfo->tm_mday);
+
+    // these should fail and not go to the keystore
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_negative@rnp", "-1"), 0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_unrecognized_1@rnp", "1z"), 0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_unrecognized_2@rnp", "now"), 0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_unrecognized_3@rnp", "00000-01-01"),
+                         0);
+    assert_int_not_equal(
+      key_generate(GENKEYS, "expiration_integer_overflow@rnp", "1234567890123456789"), 0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_32bit_overflow@rnp", "4294967296"),
+                         0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_overflow_day@rnp", "2037-02-29"),
+                         0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_overflow_month@rnp", "2037-13-01"),
+                         0);
+    if (!rnp_y2k38_warning(y2k38time)) {
+        assert_int_not_equal(
+          key_generate(GENKEYS, "expiration_overflow_year@rnp", "2337-01-01"), 0);
+    }
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_underflow_day@rnp", "2037-02-00"),
+                         0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_underflow_month@rnp", "2037-00-01"),
+                         0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_underflow_year@rnp", "1800-01-01"),
+                         0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_overflow@rnp", "200y"), 0);
+    assert_int_not_equal(key_generate(GENKEYS, "expiration_past@rnp", "2021-01-01"), 0);
+
+    // these should pass and go to the keystore -- 15 primary keys and 15 subkeys
```

Comment:
actually 17 primary keys and 17 subkeys in the final version

Reply:
- thanks, will fix via PR #1577